File: gateway/app/services/snowflake.py
import threading
import time


class SnowflakeGenerator:
    def __init__(self, worker_id: int, epoch_ms: int) -> None:
        self._worker_id = worker_id & 0x03FF
        self._epoch_ms = epoch_ms
        self._lock = threading.Lock()
        self._last_ts = -1
        self._sequence = 0

    def next_id(self) -> int:
        with self._lock:
            now = self._current_time_ms()
            if now < self._last_ts:
                now = self._last_ts
            if now == self._last_ts:
                self._sequence = (self._sequence + 1) & 0x0FFF
                if self._sequence == 0:
                    now = self._wait_next_ms(self._last_ts)
            else:
                self._sequence = 0
            self._last_ts = now
            ts_part = now - self._epoch_ms
            return (ts_part << 22) | (self._worker_id << 12) | self._sequence

    def next_id_str(self) -> str:
        return str(self.next_id())

    @staticmethod
    def _current_time_ms() -> int:
        return int(time.time() * 1000)

    def _wait_next_ms(self, last_ts: int) -> int:
        now = self._current_time_ms()
        while now <= last_ts:
            time.sleep(0)
            now = self._current_time_ms()
        return now

File: rust_gateway/src/services/snowflake.rs
use std::sync::Mutex;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

pub(crate) struct SnowflakeGenerator {
    inner: Mutex<SnowflakeState>,
}

struct SnowflakeState {
    epoch_ms: i64,
    worker_id: u16,
    last_ts: i64,
    sequence: u16,
}

impl SnowflakeGenerator {
    pub(crate) fn new(worker_id: u16, epoch_ms: i64) -> Self {
        Self {
            inner: Mutex::new(SnowflakeState {
                epoch_ms,
                worker_id,
                last_ts: -1,
                sequence: 0,
            }),
        }
    }

    pub(crate) fn next_id(&self) -> u64 {
        let mut state = self.inner.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
        let mut now = current_time_ms();
        if now < state.last_ts {
            now = state.last_ts;
        }
        if now == state.last_ts {
            state.sequence = (state.sequence + 1) & 0x0fff;
            if state.sequence == 0 {
                now = wait_next_ms(state.last_ts);
            }
        } else {
            state.sequence = 0;
        }
        state.last_ts = now;
        let ts_part = (now - state.epoch_ms) as u64;
        (ts_part << 22) | ((state.worker_id as u64) << 12) | (state.sequence as u64)
    }

    pub(crate) fn next_id_string(&self) -> String {
        self.next_id().to_string()
    }
}

fn current_time_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_else(|_| Duration::from_secs(0))
        .as_millis() as i64
}

fn wait_next_ms(last_ts: i64) -> i64 {
    let mut now = current_time_ms();
    while now <= last_ts {
        std::thread::yield_now();
        now = current_time_ms();
    }
    now
}
